[ONNX] Update typing and error messages in symbolic_helper

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83008
* __->__ #83007
* #83006

### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

justinchuby committed:
[ONNX] Update typing and error messages in symbolic_helper

### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

BowenBao reviewed (commented):
Great start to improving error and warning messages!

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -374,96 +419,104 @@ def _if_scalar_type_as(g: _C.Graph, self, tensor):
+    if not _is_list(x):
+        return False
+    x_type = x.type()
+    x_type = typing.cast(_C.ListType, x_type)
```
x.type() is annotated to be JitType or something which is the base type for ListType. I haven't found a good way to annotate this polymorphic return types. cast is needed to tell mypy that it is ListType and thus carry more methods we can use on this object. I think mypy is still going to complain about `x_type: _C.ListType = x.type()` since it would be an implicit cast ([which can only happen for Any)](https://adamj.eu/tech/2021/07/06/python-type-hints-how-to-use-typing-cast/). I suppose mypy doesn't understand `_is_list`? typing.cast is a no-op in run time, but I agree it can be noisy

BowenBao replied:
I wonder if this works
```python
def _as_list_type(type: _C.JitType) -> Optional[_C.ListType]:
    if isinstance(type, _C.ListType):
        return type
    return None
```
```python
x_type = _as_list_type(x.type())
if not x_type:
    return False
element_type = str(x_type.getElementType())
```
If it does it would appear `_as_list_type` is better than `_is_list` for mypy's sake.

justinchuby replied:
I updated to use `_as_list_type`. ptal

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -374,96 +419,104 @@ def _if_scalar_type_as(g: _C.Graph, self, tensor):
-        # Catch a more specific exception instead.
-        pass
-    return None
+def _get_tensor_dim_size(x: _C.Value, dim: int):
```
good catch thanks. Would it be Optional[int]?

justinchuby replied:
Changed to Optional[int]

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -77,69 +80,104 @@
+
+
+def _maybe_get_const(value: _C.Value, descriptor: _ValueDescriptor):
+    # FIXME(justinchuby): Can we get prim::Constant?
```
Thanks, I added this as a comment

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -77,69 +80,104 @@
-            raise RuntimeError(
-                "ONNX symbolic doesn't know to interpret ListConstruct node"
+            raise errors.SymbolicValueError(
+                f"ONNX symbolic does not know to unpack the ListConstruct node that "
```
Done

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -77,69 +80,104 @@
+                element_node = v.node()
+                if element_node.kind() != "onnx::Constant":
+                    raise errors.SymbolicValueError(
+                        f"Failed to export an ONNX attribute '{element_node.kind()}' "
```
Done

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -482,7 +483,7 @@ def _get_tensor_rank(x: _C.Value) -> Optional[int]:
 
 
-def _get_tensor_sizes(x: _C.Value, allow_nonstatic=True):
+def _get_tensor_sizes(x: _C.Value, allow_nonstatic: bool = True):
```
The polymorphic return type (Optional[List[Optional[int]]]) causes many troubles in callers when mypy does its analysis so I am skipping it

justinchuby committed:
Update on "[ONNX] Update typing and error messages in symbolic_helper"


### Description

- Clearer error messages with more context
-   Created `SymbolicValueError` which adds context of the value to the error message
- Type annotation

example error message:

```
torch.onnx.errors.SymbolicValueError: ONNX symbolic does not understand the Constant node '%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
' specified with descriptor 'is'.  [Caused by the value '1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Constant'.]

    Inputs:
        Empty
    Outputs:
        #0: 1 defined in (%1 : Long(2, strides=[1], device=cpu) = onnx::Constant[value= 3  3 [ CPULongType{2} ]]()
    )  (type 'Tensor')
```

### Issue

- #77316 (Runtime error during symbolic conversion)

### Testing

Unit tested

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge
